content/outcomes/secure-software-supply-chain/building-a-container.md

```diff
@@ -0,0 +1,91 @@
+---
+title: Building a Container
+weight: 2
+layout: single
+---
+
+There's many options that you have when deciding how you want to build the container for your code. From writing a Dockerfile by hand to automated build services, finding the right build method is an important first step. Let's take a look at a few options that exist do see the advantages and disadvantages of each!
+
+## Dockerfile
+
+If you're already familiar with building a container image, this is probably where you got your start. A Dockerfile defines a set of steps to build you container, as well as if it should use another container image as a base. This allows us to reuse more generic container images that may contain things such as the appropriate application runtimes and dependencies. Consider this basic Dockerfile:
+
+```dockerfile
+FROM ubuntu:18.04
+COPY . /app
+RUN make /app
+ENTRYPOINT /app/my-apo
+```
+
+This Dockerfile tells the builder to use the `ubuntu:18.04` container image as a base and to copy the current directory from the local filesystem to `/app` within the container that we're building. The `RUN` command then runs the command to build the application. Finally, the `ENTRYPOINT` instruction defines what to run when the container is started. In this case, when the container is started, the executable at `/app/my-app` is ran.
+
+With that basic understanding, let's look at a more specific example. In this case, this Dockerfile is used to run a Spring Boot application:
+
+```dockerfile
+FROM openjdk:8-jdk-alpine
+EXPOSE 8080
+ARG JAR_FILE=target/my-application.jar
+ADD ${JAR_FILE} app.jar
+ENTRYPOINT ["java","-jar","/app.jar"]
+```
+
+Let's start with the two new instructions added here. Since our container will be running a web application running on port 8080, we can inform the container runtime of this by using the `EXPOSE` instruction. Additionally, from the `ADD` instruction, you might have noticed that this Dockerfile is expecting a prebuilt JAR. If you're a Java developer, you may know that different build tools, such as Maven and Gradle, build their applications differently. The `ARG` instruction defines a variable that can be used in the Dockerfile, with the option to override it when building the container. In this case, we set the default value of `JAR_FILE` to `target/my-application.jar` with the assumption that we're building this application Maven. If this is the case, the container could then just be built with the following command:
```

Comment:
building this application Maven => building this application *with* Maven outside the context of the Docker build process.